services/gossip/topic_block_sync.go

```diff
@@ -1,20 +1,200 @@
+		RecipientMode:       gossipmessages.RECIPIENT_LIST_MODE_LIST,
+		RecipientPublicKeys: []primitives.Ed25519PublicKey{input.RecipientPublicKey},
+	}).Build()
+
+	if input.Message.SignedChunkRange == nil {
+		return nil, errors.Errorf("cannot encode BlockSyncRequestMessage", log.Stringable("message", input.Message))
+	}
+	payloads := [][]byte{header.Raw(), input.Message.SignedChunkRange.Raw(), input.Message.Sender.Raw()}
+
+	return nil, s.transport.Send(&adapter.TransportData{
+		SenderPublicKey:     s.config.NodePublicKey(),
+		RecipientMode:       gossipmessages.RECIPIENT_LIST_MODE_LIST,
+		RecipientPublicKeys: []primitives.Ed25519PublicKey{input.RecipientPublicKey},
+		Payloads:            payloads,
+	})
+}
+
+func (s *service) receivedBlockSyncRequest(header *gossipmessages.Header, payloads [][]byte) {
+	if len(payloads) < 2 {
+		return
+	}
+	chunkRange := gossipmessages.BlockSyncRangeReader(payloads[0])
+	senderSignature := gossipmessages.SenderSignatureReader(payloads[1])
+
+	for _, l := range s.blockSyncHandlers {
+		l.HandleBlockSyncRequest(&gossiptopics.BlockSyncRequestInput{
+			Message: &gossipmessages.BlockSyncRequestMessage{
+				SignedChunkRange: chunkRange,
+				Sender:           senderSignature,
+			},
+		})
+	}
 }
+
 func (s *service) SendBlockSyncResponse(input *gossiptopics.BlockSyncResponseInput) (*gossiptopics.EmptyOutput, error) {
-	panic("Not implemented")
+	header := (&gossipmessages.HeaderBuilder{
+		Topic:               gossipmessages.HEADER_TOPIC_BLOCK_SYNC,
+		BlockSync:           gossipmessages.BLOCK_SYNC_RESPONSE,
+		RecipientMode:       gossipmessages.RECIPIENT_LIST_MODE_LIST,
+		RecipientPublicKeys: []primitives.Ed25519PublicKey{input.RecipientPublicKey},
+	}).Build()
+
+	if input.Message.SignedChunkRange == nil || len(input.Message.BlockPairs) == 0 {
+		return nil, errors.Errorf("cannot encode BlockSyncResponseMessage", log.Stringable("message", input.Message))
+	}
+	payloads := [][]byte{header.Raw(), input.Message.SignedChunkRange.Raw(), input.Message.Sender.Raw()}
+
+	for _, blocks := range input.Message.BlockPairs {
+		blockPairPayloads, err := encodeBlockPair(blocks)
+		if err != nil {
+			return nil, err
+		}
+		payloads = append(payloads, blockPairPayloads...)
+	}
+
+	return nil, s.transport.Send(&adapter.TransportData{
+		SenderPublicKey:     s.config.NodePublicKey(),
+		RecipientMode:       gossipmessages.RECIPIENT_LIST_MODE_LIST,
+		RecipientPublicKeys: []primitives.Ed25519PublicKey{input.RecipientPublicKey},
+		Payloads:            payloads,
+	})
+}
+
+func (s *service) receivedBlockSyncResponse(header *gossipmessages.Header, payloads [][]byte) {
+	if len(payloads) < 3 {
+		return
+	}
+	chunkRange := gossipmessages.BlockSyncRangeReader(payloads[0])
+	senderSignature := gossipmessages.SenderSignatureReader(payloads[1])
+	blockPairPayloads := payloads[2:]
+
+	var blocks []*protocol.BlockPairContainer
+
+	for i := 0; i < len(blockPairPayloads); i += 8 {
```

Comment:
ahhhh, you can't use 8 as a hard coded value

this number depends on the number of transactions, receipts and state diffs in the block!

you have to have a dynamic calculation

Reply:
Fixed